src/libcore/array.rs

```diff
@@ -62,6 +64,111 @@ unsafe impl<T, A: Unsize<[T]>> FixedSizeArray<T> for A {
     }
 }
 
+/// An iterator that moves out of an array.
+#[derive(Debug)]
+pub struct IntoIter<T, A: FixedSizeArray<T>> {
+    // Invariants: index <= index_back <= array.len()
+    // Only values in array[index..index_back] are alive at any given time.
+    // Values from array[..index] and array[index_back..] are already moved/dropped.
+    array: Option<A>,
+    index: usize,
+    index_back: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T, A: FixedSizeArray<T>> Drop for IntoIter<T, A> {
+    #[inline]
+    fn drop(&mut self) {
+        // Drop values that are still alive.
+        if let Some(array) = self.array.as_mut() {
+            let slice = array.as_mut_slice();
+            for p in &mut slice[self.index..self.index_back] {
+                unsafe { ptr::drop_in_place(p); }
+            }
+        }
+
+        // Prevent the array as a whole from dropping.
+        unsafe { ptr::write(&mut self.array, None); }
+    }
+}
+
+impl<T, A: FixedSizeArray<T>> Iterator for IntoIter<T, A> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        if self.len() > 0 {
+            if let Some(array) = self.array.as_ref() {
```

Comment:
Right, it should never be `None`.  I suppose `unwrap` expresses that better, but it will introduce a panic into codegen, whereas this currently would act like just the end of iteration.  I don't mind changing this though.